### PR TITLE
ASMs2Dmx: add support for a separate geometry basis

### PR DESCRIPTION
**Pull request by akva2**
Ths enables a separate geometry in ASMs2Dmx.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=236 ifem-navierstokes=191 ifem-boussinesq=95 please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=240 please

File: src/ASM/ASMs2D.C [hunk 1]
```diff
@@ -1365,6 +1365,8 @@ bool ASMs2D::getElementCoordinates (Matrix& X, int iel) const
 #endif
 
   const Go::SplineSurface* geo = this->getBasis(ASM::GEOMETRY_BASIS);
+  if (forceItg)
+    geo = surf;
```
Comment (kmokstad) [inline src/ASM/ASMs2D.C:1369]: Isn't it better with `const Go::SplineSurface* geo = forceItg ? surf : this->getBasis(ASM::GEOMETRY_BASIS);` such that `getBasis()` is invoked only when needed?

File: src/ASM/ASMs3D.C [hunk 2]
```diff
@@ -1622,7 +1622,7 @@ Vec3 ASMs3D::getCoord (size_t inod) const
 }
 
 
-bool ASMs3D::getElementCoordinates (Matrix& X, int iel) const
+bool ASMs3D::getElementCoordinates (Matrix& X, int iel, bool) const
```
Comment (kmokstad) [inline src/ASM/ASMs3D.C:1625]: Wouldn't you need the same `forceItg` stuff here in the 3D?

File: src/ASM/ASMs2D.C [hunk 3]
```diff
@@ -1587,12 +1587,16 @@ bool ASMs2D::getParameterDomain (Real2DMat& u, IntVec* corners) const
 
 
 const Vector& ASMs2D::getGaussPointParameters (Matrix& uGP, int dir, int nGauss,
-					       const double* xi) const
+                                               const double* xi,
+                                              const Go::SplineSurface* spline) const
```
Comment (kmokstad) [inline src/ASM/ASMs2D.C:1591]: incorrect indentation

File: src/ASM/ASMs2D.h [hunk 4]
```diff
@@ -240,8 +240,9 @@ class ASMs2D : public ASMstruct, public ASM2D
   //! \brief Returns a matrix with nodal coordinates for an element.
   //! \param[in] iel Element index
   //! \param[out] X 3\f$\times\f$n-matrix, where \a n is the number of nodes
+  //! \param[in] forceItg If true return integration basis element coordinates
   //! in one element
```
Comment (kmokstad) [inline src/ASM/ASMs2D.h:244]: The added doxy line is misplaced, as the "in one element" belongs to the X-parameter description.

**Review by kmokstad — APPROVED** [review]
Conditionally approved, with some questions/suggestions

File: src/ASM/ASMs2D.C [hunk 5]
```diff
@@ -1359,15 +1359,24 @@ bool ASMs2D::getElementCoordinates (Matrix& X, int iel) const
...
 
-  X.resize(nsd,surf->order_u()*surf->order_v());
+  const Go::SplineSurface* geo = this->getBasis(ASM::GEOMETRY_BASIS);
 
   int lnod0 = this->getFirstItgElmNode();
-  RealArray::const_iterator cit = surf->coefs_begin();
+  if (geo != surf) {
+    const int node = MNPC[iel-1][lnod0];
+    double u = surf->basis_u().getKnots()[nodeInd[node].I + surf->order_u() - 1];
+    double v = surf->basis_v().getKnots()[nodeInd[node].J + surf->order_v() - 1];
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2D.C]: prettier in any case so done.

File: src/ASM/ASMs2D.C [hunk 6]
```diff
@@ -1384,6 +1393,37 @@ bool ASMs2D::getElementCoordinates (Matrix& X, int iel) const
...
+    nj = geo->basis_v().knotInterval(v) - geo->order_v() + 1;
+  }
+
+  RealArray::const_iterator cit = geo->coefs_begin();
+  for (size_t n = 0; n < X.cols(); n++)
+  {
+    const int iu = n % geo->order_u();
+    const int iv = n / geo->order_u();
+    const int ip = (ni + iu + (nj + iv)*geo->numCoefs_u())*geo->dimension();
+    if (ip < 0) return false;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2D.C]: yeah, copy-pasta issue. i've added an appropriate check. i have to check parameters and not ni/nj as gotools just returns first / last knotinterval if we are out of bounds.

File: src/ASM/ASMs2D.C (hunk 1 above)
Comment (akva2) [reply src/ASM/ASMs2D.C]: sure, i guess we can eat the long line for the minor optim.

File: src/ASM/ASMs3D.C (hunk 2 above)
Comment (akva2) [reply src/ASM/ASMs3D.C:1625]: separate PR.

File: src/ASM/ASMs2D.C (hunk 3 above)
Comment (akva2) [reply src/ASM/ASMs2D.C]: fixed.

File: src/ASM/ASMs2D.h (hunk 4 above)
Comment (akva2) [reply src/ASM/ASMs2D.h:243]: fixed.

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=240 please

**Pull request merged**